File: stego/pipeline.py
"""Format-independent payload pipeline (concatenated code).

A carrier exposes a flat array of 1-bit *slots* (low bit-planes of its samples). Layout:

    payload --frame(name+len+crc)--> [zlib?] --> [AES-GCM?] --> fountain symbols
    each symbol --> + inner CRC16 --> Reed-Solomon parity --> wire bytes
    slots = [ self-checking numeric HEADER, replicated & spread across the whole carrier ]
            [ RS-wrapped fountain symbols, spread evenly into cells across the carrier   ]

Redundancy is two-layered and tunable:
  * inner Reed-Solomon repairs scattered bit/byte flips *within* a symbol (light noise);
  * each symbol occupies a contiguous cell, and cells are spread across the whole carrier, so a
    localized edit/overwrite erases a *proportional* set of whole symbols, which the outer
    fountain (LT) rebuilds from any K'~K survivors. A payload embedded at R× tolerates losing
    ~(1-1/R) of the carrier; embedding at "fill" maximizes R and thus damage tolerance.
The header is replicated R_HEADER× (each copy self-checked by CRC) and spread, so it is read by
majority vote (beats scattered noise) with a single-intact-copy fallback (beats huge overwrites).
Decode is fail-safe: header CRC, per-symbol CRC, and a final payload CRC must all pass.
"""
import math, struct, zlib
import numpy as np
from .bitio import bytes_to_bits, bits_to_bytes, crc16
from .fountain import Fountain, blocks_of, decode as lt_decode
from . import crypto

class PipelineError(Exception): pass

MAGIC       = b"MDX1"
VERSION     = 1
R_HEADER    = 11
_HDR_FMT    = ">4sBBHIIIIBBI"                     # magic,ver,flags,block_size,K,seed,coded_len,n_sym,n_lsb,rs_nsym,crc
HEADER_BYTES = struct.calcsize(_HDR_FMT)          # 30
FLAG_ZLIB, FLAG_AES = 1, 2
DEFAULT_BLOCK, DEFAULT_RS = 64, 20

def _rscodec(nsym):
    if nsym <= 0: return None
    from reedsolo import RSCodec
    return RSCodec(nsym)

def _rs_decode(rsc, chunk):
    out = rsc.decode(chunk)
    return bytes(out[0] if isinstance(out, tuple) else out)

# ---------- frame (filename + integrity) ----------
def _frame(payload, filename):
    name = (filename or "payload.bin").encode("utf-8")[:65535]
    return struct.pack(">HII", len(name), len(payload), zlib.crc32(payload) & 0xffffffff) + name + payload

def _unframe(buf):
    if len(buf) < 10: raise PipelineError("frame too short")
    name_len, data_len, crc = struct.unpack(">HII", buf[:10])
    name = buf[10:10 + name_len].decode("utf-8", "replace")
    data = buf[10 + name_len:10 + name_len + data_len]
    if len(data) != data_len: raise PipelineError("payload truncated")
    if (zlib.crc32(data) & 0xffffffff) != crc: raise PipelineError("payload CRC-32 mismatch (corrupted beyond recovery)")
    return name, data

# ---------- compress + encrypt middle layer ----------
def pack(payload, filename="payload.bin", password=None):
    frame = _frame(payload, filename); flags = 0
    comp = zlib.compress(frame, 9); coded = frame
    if len(comp) < len(frame): coded = comp; flags |= FLAG_ZLIB
    if password: coded = crypto.encrypt(coded, password); flags |= FLAG_AES
    return coded, flags

def unpack(coded, flags, password=None):
    if flags & FLAG_AES:  coded = crypto.decrypt(coded, password)
    if flags & FLAG_ZLIB: coded = zlib.decompress(coded)
    return _unframe(coded)

# ---------- header ----------
def _make_header(flags, block_size, K, seed, coded_len, n_sym, n_lsb, rs_nsym):
    h = struct.pack(_HDR_FMT, MAGIC, VERSION, flags, block_size, K, seed, coded_len, n_sym, n_lsb, rs_nsym, 0)
    return h[:-4] + struct.pack(">I", zlib.crc32(h[:-4]) & 0xffffffff)

def _read_header_bytes(hbytes):
    if hbytes[:4] != MAGIC: return None
    _, ver, flags, block_size, K, seed, coded_len, n_sym, n_lsb, rs_nsym, crc = struct.unpack(_HDR_FMT, hbytes)
    if (zlib.crc32(hbytes[:-4]) & 0xffffffff) != crc: return None
    return dict(version=ver, flags=flags, block_size=block_size, K=K, seed=seed,
                coded_len=coded_len, n_sym=n_sym, n_lsb=n_lsb, rs_nsym=rs_nsym)

def _step_bytes(block_size, rs_nsym):
    # RS path carries data(B)+inner CRC16(2) under RS parity (CRC catches RS mis-corrections,
    # e.g. an overwrite-to-silence is a valid all-zero RS codeword and must be rejected).
    return (block_size + 2 + rs_nsym) if rs_nsym > 0 else (2 + block_size)

# ---------- deterministic slot layout (shared with the JS app) ----------
def _layout(cap):
    """(header_positions, body_positions). Header spread by a wide stride into R_HEADER full,
       self-checking copies; the body takes the remaining slots in order (then cell-spread)."""
    nbits = HEADER_BYTES * 8
    header_total = R_HEADER * nbits
    if cap < header_total + 1:
        return None
    stride = cap // header_total
    hpos = np.arange(header_total, dtype=np.int64) * stride      # copy-major: copy c occupies slice c
    mask = np.ones(cap, dtype=bool); mask[hpos] = False
    return hpos, np.flatnonzero(mask)

def _cell(i, n_fit, n_sym):
    return (i * n_fit) // n_sym                                   # spread n_sym symbols across n_fit cells

# ---------- slot-level embed / extract ----------
def embed_slots(slot_bits, payload, filename, password=None, block_size=DEFAULT_BLOCK,
                n_lsb=1, redundancy=None, rs_nsym=DEFAULT_RS):
    if rs_nsym > 0 and block_size + 2 + rs_nsym > 255:
        raise PipelineError("block_size + 2 + rs_nsym must be <= 255 (Reed-Solomon GF(256) limit)")
    coded, flags = pack(payload, filename, password)
    seed = zlib.crc32(coded) & 0xffffffff
    blocks, K = blocks_of(coded, block_size)
    fnt = Fountain(K, block_size, seed); rsc = _rscodec(rs_nsym)
    cap = len(slot_bits)
    lay = _layout(cap)
    if lay is None: raise PipelineError("carrier far too small")
    hpos, body = lay
    step = _step_bytes(block_size, rs_nsym); symbol_slots = step * 8
    n_fit = len(body) // symbol_slots
    if n_fit < K:
        raise PipelineError("carrier too small: fits %d symbols but payload needs >= %d" % (n_fit, K))
    n_sym = n_fit if redundancy is None else int(min(n_fit, max(K, math.ceil(K * redundancy))))
    out = slot_bits.copy()
    header = _make_header(flags, block_size, K, seed, len(coded), n_sym, n_lsb, rs_nsym)
    out[hpos] = np.tile(bytes_to_bits(header), R_HEADER)
    for i in range(n_sym):
        sym = fnt.encode_symbol(blocks, i).tobytes()
        wire = bytes(rsc.encode(sym + struct.pack(">H", crc16(sym)))) if rsc else (struct.pack(">H", crc16(sym)) + sym)
        c = _cell(i, n_fit, n_sym)
        out[body[c * symbol_slots:(c + 1) * symbol_slots]] = bytes_to_bits(wire)
    info = dict(K=K, block_size=block_size, rs_nsym=rs_nsym, n_symbols=n_sym, redundancy=round(n_sym / K, 2),
                coded_len=len(coded), flags=flags, capacity_slots=cap, n_lsb=n_lsb,
                tolerates_symbol_loss="%.0f%%" % (100 * (1 - K / n_sym)))
    return out, info

def extract_slots(slot_bits, hdr):
    cap = len(slot_bits)
    hpos, body = _layout(cap)
    B, K, seed, coded_len, n_sym, rs_nsym = (hdr["block_size"], hdr["K"], hdr["seed"],
                                             hdr["coded_len"], hdr["n_sym"], hdr["rs_nsym"])
    step = _step_bytes(B, rs_nsym); symbol_slots = step * 8
    n_fit = len(body) // symbol_slots
    fnt = Fountain(K, B, seed); rsc = _rscodec(rs_nsym)
    symbols, recovered = {}, 0
    for i in range(n_sym):
        c = _cell(i, n_fit, n_sym)
        wire = bits_to_bytes(slot_bits[body[c * symbol_slots:(c + 1) * symbol_slots]])
        if len(wire) < step: continue
        if rsc is not None:
            try: inner = _rs_decode(rsc, wire)
            except Exception: continue
            data = inner[:B]; (crc,) = struct.unpack(">H", inner[B:B + 2])
            if crc16(data) == crc: symbols[i] = data; recovered += 1
        else:
            (crc,) = struct.unpack(">H", wire[:2]); data = wire[2:]
            if crc16(data) == crc: symbols[i] = data; recovered += 1
    coded_padded = lt_decode(symbols, fnt)
    if coded_padded is None:
        raise PipelineError("too many symbols lost — payload unrecoverable (%d/%d symbols usable)" % (recovered, n_sym))
    return coded_padded[:coded_len]

def read_header(slot_bits):
    lay = _layout(len(slot_bits))
    if lay is None: return None
    hpos, _ = lay; nbits = HEADER_BYTES * 8
    copies = slot_bits[hpos].reshape(R_HEADER, nbits)
    maj = _read_header_bytes(bits_to_bytes((copies.sum(axis=0) * 2 > R_HEADER).astype(np.uint8)))
    if maj: return maj                                           # majority vote: beats scattered noise
    for c in range(R_HEADER):                                    # fallback: any one intact copy beats huge overwrites
        h = _read_header_bytes(bits_to_bytes(copies[c].astype(np.uint8)))
        if h: return h
    return None

# ---------- what carriers call upward ----------
def embed(carrier, payload, filename="payload.bin", password=None, block_size=DEFAULT_BLOCK,
          n_lsb=1, redundancy=None, rs_nsym=DEFAULT_RS):
    bits = carrier.get_bits(n_lsb)
    new_bits, info = embed_slots(bits, payload, filename, password, block_size, n_lsb, redundancy, rs_nsym)
    carrier.set_bits(new_bits, n_lsb)
    return info

def extract(carrier, password=None, n_lsb_candidates=(1, 2, 3, 4)):
    for n_lsb in n_lsb_candidates:
        bits = carrier.get_bits(n_lsb)
        hdr = read_header(bits)
        if hdr is None or hdr["n_lsb"] != n_lsb:
            continue
        coded = extract_slots(bits, hdr)
        name, data = unpack(coded, hdr["flags"], password)
        return name, data, dict(hdr, n_lsb=n_lsb)
    raise PipelineError("no MDX1 header found (not a stego carrier, or too corrupted to read)")

File: web/src/stego/pipeline.js
// Format-independent payload pipeline, JS port of stego/pipeline.py.
// Uses rs_nsym=0 mode (no inner Reed-Solomon) so we don't need to port the RS library;
// inner integrity is CRC-16/CCITT and fountain handles outer redundancy. Lossless audio is
// bit-exact anyway, so RS isn't needed for the lossless cross-compat path.

import { bytesToBits, bitsToBytes, crc16, crc32 } from "./bitio.js";
import { Fountain, ltDecode, blocksOf } from "./fountain.js";
import { deflate, inflate } from "./zlib.js";

export class PipelineError extends Error { constructor(m) { super(m); this.name = "PipelineError"; } }

const MAGIC      = new TextEncoder().encode("MDX1");
const VERSION    = 1;
const R_HEADER   = 11;
export const HEADER_BYTES   = 30;    // matches Python _HDR_FMT
export const DEFAULT_BLOCK  = 64;
export const FLAG_ZLIB = 1;

// ---------- byte-level helpers ----------
function dvWrite(buf, off, fmt, ...vals) {
  const v = new DataView(buf.buffer, buf.byteOffset + off);
  let p = 0;
  for (let i = 0; i < fmt.length; i++) {
    const c = fmt[i];
    if      (c === "B") { v.setUint8(p, vals[i]);             p += 1; }
    else if (c === "H") { v.setUint16(p, vals[i], false);     p += 2; }
    else if (c === "I") { v.setUint32(p, vals[i] >>> 0, false); p += 4; }
    else if (c === "s") { buf.set(vals[i], off + p);          p += vals[i].length; }
    else throw new Error("bad fmt " + c);
  }
}
function dvRead(buf, off, fmt) {
  const v = new DataView(buf.buffer, buf.byteOffset + off);
  const out = []; let p = 0;
  for (const c of fmt) {
    if      (c === "B") { out.push(v.getUint8(p)); p += 1; }
    else if (c === "H") { out.push(v.getUint16(p, false)); p += 2; }
    else if (c === "I") { out.push(v.getUint32(p, false) >>> 0); p += 4; }
    else throw new Error("bad fmt " + c);
  }
  return out;
}
function concatU8(...arrs) {
  let n = 0; for (const a of arrs) n += a.length;
  const out = new Uint8Array(n); let p = 0;
  for (const a of arrs) { out.set(a, p); p += a.length; }
  return out;
}

// ---------- frame (filename + integrity) ----------
function frame(payload, filename) {
  const name = new TextEncoder().encode(filename || "payload.bin").slice(0, 65535);
  const head = new Uint8Array(2 + 4 + 4);
  const v = new DataView(head.buffer);
  v.setUint16(0, name.length, false);
  v.setUint32(2, payload.length, false);
  v.setUint32(6, crc32(payload), false);
  return concatU8(head, name, payload);
}
function unframe(buf) {
  if (buf.length < 10) throw new PipelineError("frame too short");
  const v = new DataView(buf.buffer, buf.byteOffset);
  const nameLen = v.getUint16(0, false);
  const dataLen = v.getUint32(2, false);
  const crc     = v.getUint32(6, false) >>> 0;
  const name = new TextDecoder().decode(buf.slice(10, 10 + nameLen));
  const data = buf.slice(10 + nameLen, 10 + nameLen + dataLen);
  if (data.length !== dataLen) throw new PipelineError("payload truncated");
  if (crc32(data) !== crc) throw new PipelineError("payload CRC-32 mismatch (corrupted beyond recovery)");
  return { name, data };
}

// ---------- pack / unpack (compression layer; no AES in browser yet) ----------
export function pack(payload, filename = "payload.bin") {
  const fr = frame(payload, filename);
  let flags = 0;
  let coded = fr;
  const comp = deflate(fr);
  if (comp.length < fr.length) { coded = comp; flags |= FLAG_ZLIB; }
  return { coded, flags };
}
export function unpack(coded, flags) {
  let buf = coded;
  if (flags & FLAG_ZLIB) {
    try { buf = inflate(coded); }
    catch (e) { throw new PipelineError("decompression failed (payload corrupted): " + e.message); }
  }
  return unframe(buf);
}

// ---------- header (30 bytes; same layout as Python _HDR_FMT) ----------
//   magic(4) ver(1) flags(1) block_size(2) K(4) seed(4) coded_len(4) n_sym(4) n_lsb(1) rs_nsym(1) crc(4)
function makeHeader(flags, blockSize, K, seed, codedLen, nSym, nLsb, rsNsym) {
  const h = new Uint8Array(HEADER_BYTES);
  dvWrite(h, 0, "sBBHIIIIBB", MAGIC, VERSION, flags, blockSize, K, seed, codedLen, nSym, nLsb, rsNsym);
  const c = crc32(h.slice(0, HEADER_BYTES - 4));
  new DataView(h.buffer).setUint32(HEADER_BYTES - 4, c, false);
  return h;
}
function readHeaderBytes(h) {
  for (let i = 0; i < 4; i++) if (h[i] !== MAGIC[i]) return null;
  const [, ver, flags, blockSize, K, seed, codedLen, nSym, nLsb, rsNsym, crc] =
        dvRead(h, 0, "IBBHIIIIBBI");
  void [ver];                                              // version read but unused
  if (crc32(h.slice(0, HEADER_BYTES - 4)) !== crc) return null;
  return { version: ver, flags, blockSize, K, seed, codedLen, nSym, nLsb, rsNsym };
}

function stepBytes(blockSize, rsNsym) {
  return rsNsym > 0 ? (blockSize + 2 + rsNsym) : (2 + blockSize);
}

// Same layout function as stego/pipeline.py:_layout
function layout(cap) {
  const nbits = HEADER_BYTES * 8;
  const headerTotal = R_HEADER * nbits;
  if (cap < headerTotal + 1) return null;
  const stride = Math.floor(cap / headerTotal);
  const hpos = new Int32Array(headerTotal);
  for (let i = 0; i < headerTotal; i++) hpos[i] = i * stride;
  const mask = new Uint8Array(cap).fill(1);
  for (let i = 0; i < headerTotal; i++) mask[hpos[i]] = 0;
  const body = []; for (let i = 0; i < cap; i++) if (mask[i]) body.push(i);
  return { hpos, body: new Int32Array(body) };
}

const cell = (i, nFit, nSym) => Math.floor((i * nFit) / nSym);

// ---------- the four functions a carrier needs ----------
export function embedSlots(slotBits, payload, filename, { blockSize = DEFAULT_BLOCK, nLsb = 1, redundancy = null } = {}) {
  const { coded, flags } = pack(payload, filename);
  const seed = crc32(coded);
  const { blocks, K } = blocksOf(coded, blockSize);
  const fnt = new Fountain(K, blockSize, seed);
  const rsNsym = 0;                                        // browser uses no inner RS
  const cap = slotBits.length;
  const lay = layout(cap);
  if (lay === null) throw new PipelineError("carrier far too small");
  const { hpos, body } = lay;
  const step = stepBytes(blockSize, rsNsym), symSlots = step * 8;
  const nFit = Math.floor(body.length / symSlots);
  if (nFit < K) throw new PipelineError(`carrier too small: fits ${nFit} symbols but payload needs >= ${K}`);
  const nSym = redundancy == null ? nFit : Math.min(nFit, Math.max(K, Math.ceil(K * redundancy)));
  const out = new Uint8Array(slotBits);
  const header = makeHeader(flags, blockSize, K, seed, coded.length, nSym, nLsb, rsNsym);
  const hbits = bytesToBits(header);
  for (let c = 0; c < R_HEADER; c++)
    for (let b = 0; b < hbits.length; b++) out[hpos[c * hbits.length + b]] = hbits[b];
  for (let i = 0; i < nSym; i++) {
    const sym = fnt.encodeSymbol(blocks, i);
    const wire = new Uint8Array(2 + sym.length);
    new DataView(wire.buffer).setUint16(0, crc16(sym), false);
    wire.set(sym, 2);
    const wbits = bytesToBits(wire);
    const c = cell(i, nFit, nSym);
    for (let b = 0; b < wbits.length; b++) out[body[c * symSlots + b]] = wbits[b];
  }
  return {
    bits: out,
    info: {
      K, blockSize, rsNsym, nSymbols: nSym, redundancy: Math.round(100 * nSym / K) / 100,
      codedLen: coded.length, flags, capacitySlots: cap, nLsb,
      tolerates_symbol_loss: `${Math.round(100 * (1 - K / nSym))}%`
    }
  };
}

export function readHeader(slotBits) {
  const lay = layout(slotBits.length);
  if (lay === null) return null;
  const { hpos } = lay;
  const nbits = HEADER_BYTES * 8;
  const votes = new Int32Array(nbits);
  for (let c = 0; c < R_HEADER; c++)
    for (let b = 0; b < nbits; b++) votes[b] += slotBits[hpos[c * nbits + b]];
  const maj = new Uint8Array(nbits);
  for (let b = 0; b < nbits; b++) maj[b] = (votes[b] * 2 > R_HEADER) ? 1 : 0;
  const tryParse = (bits) => readHeaderBytes(bitsToBytes(bits));
  let hdr = tryParse(maj);
  if (hdr) return hdr;
  for (let c = 0; c < R_HEADER; c++) {                     // fallback: any single intact copy
    const copy = new Uint8Array(nbits);
    for (let b = 0; b < nbits; b++) copy[b] = slotBits[hpos[c * nbits + b]];
    hdr = tryParse(copy);
    if (hdr) return hdr;
  }
  return null;
}

export function extractSlots(slotBits, hdr) {
  const lay = layout(slotBits.length);
  const { body } = lay;
  const { blockSize: B, K, seed, codedLen, nSym, rsNsym } = hdr;
  if (rsNsym > 0) throw new PipelineError("rs_nsym>0 (Python-only path) — install reedsolo-js to extract");
  const step = stepBytes(B, rsNsym), symSlots = step * 8;
  const nFit = Math.floor(body.length / symSlots);
  const fnt = new Fountain(K, B, seed);
  const symbols = new Map();
  for (let i = 0; i < nSym; i++) {
    const c = cell(i, nFit, nSym);
    const sBits = new Uint8Array(symSlots);
    for (let b = 0; b < symSlots; b++) sBits[b] = slotBits[body[c * symSlots + b]];
    const wire = bitsToBytes(sBits);
    if (wire.length < step) continue;
    const crc = new DataView(wire.buffer, wire.byteOffset).getUint16(0, false);
    const data = wire.slice(2);
    if (crc16(data) === crc) symbols.set(i, data);
  }
  const codedPadded = ltDecode(symbols, fnt);
  if (codedPadded === null)
    throw new PipelineError(`too many symbols lost — payload unrecoverable (${symbols.size}/${nSym} usable)`);
  return codedPadded.slice(0, codedLen);
}

// ---------- public file API (carrier-agnostic) ----------
export function embed(carrier, payload, filename = "payload.bin", opts = {}) {
  const bits = carrier.getBits(opts.nLsb || 1);
  const { bits: newBits, info } = embedSlots(bits, payload, filename, opts);
  carrier.setBits(newBits, opts.nLsb || 1);
  return info;
}
export function extract(carrier, nLsbCandidates = [1, 2, 3, 4]) {
  for (const nLsb of nLsbCandidates) {
    let bits;
    try { bits = carrier.getBits(nLsb); } catch (e) { continue; }
    const hdr = readHeader(bits);
    if (hdr === null || hdr.nLsb !== nLsb) continue;
    const coded = extractSlots(bits, hdr);
    const { name, data } = unpack(coded, hdr.flags);
    return { name, data, hdr };
  }
  throw new PipelineError("no MDX1 header found (not a stego carrier, or too corrupted to read)");
}
